src/main/java/org/opensearch/neuralsearch/processor/NeuralSparseTwoPhaseProcessor.java

```diff
@@ -0,0 +1,250 @@
+                String.format(Locale.ROOT, "The two_phase_parameter.max_window_size must >= 50. Received: %n" + maxWindowSize)
+            );
+        }
+        this.maxWindowSize = maxWindowSize;
+    }
+
+    /**
+     * Process the search request of neural_sparse_two_phase_processor
+     * @param request the search request (which may have been modified by an earlier processor)
+     * @return request the search request that add the two-phase rescore query of neural sparse query.
+     */
+    @Override
+    public SearchRequest processRequest(final SearchRequest request) {
+        if (!enabled || ratio == 0f) {
+            return request;
+        }
+        QueryBuilder queryBuilder = request.source().query();
+        // Collect the nested NeuralSparseQueryBuilder in the whole query.
+        Multimap<NeuralSparseQueryBuilder, Float> queryBuilderMap;
+        queryBuilderMap = collectNeuralSparseQueryBuilder(queryBuilder, 1.0f);
+        if (queryBuilderMap.isEmpty()) {
+            return request;
+        }
+        // Make a nestedQueryBuilder which includes all the two-phase QueryBuilder.
+        QueryBuilder nestedTwoPhaseQueryBuilder = getNestedQueryBuilderFromNeuralSparseQueryBuilderMap(queryBuilderMap);
+        nestedTwoPhaseQueryBuilder.boost(getOriginQueryWeightAfterRescore(request.source()));
+        // Add it to the rescorer.
+        RescorerBuilder<QueryRescorerBuilder> twoPhaseRescorer = buildRescoreQueryBuilderForTwoPhase(nestedTwoPhaseQueryBuilder, request);
+        request.source().addRescorer(twoPhaseRescorer);
+        return request;
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    /**
+     * Based on ratio, split a Map into two map by the value.
+     *
+     * @param queryTokens    the queryTokens map, key is the token String, value is the score.
+     * @param thresholdRatio The ratio that control how tokens map be split.
+     * @return A map has two element, {[True, token map whose value above threshold],[False, token map whose value below threshold]}
+     */
+    public static Tuple<Map<String, Float>, Map<String, Float>> splitQueryTokensByRatioedMaxScoreAsThreshold(
+        final Map<String, Float> queryTokens,
+        final float thresholdRatio
+    ) {
+
+        float max = 0f;
+        for (Float value : queryTokens.values()) {
+            max = Math.max(value, max);
+        }
+        float threshold = max * thresholdRatio;
+
+        Map<Boolean, Map<String, Float>> queryTokensByScore = queryTokens.entrySet()
+            .stream()
+            .collect(
+                Collectors.partitioningBy(entry -> entry.getValue() >= threshold, Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue))
+            );
+
+        Map<String, Float> highScoreTokens = queryTokensByScore.get(Boolean.TRUE);
+        Map<String, Float> lowScoreTokens = queryTokensByScore.get(Boolean.FALSE);
+        if (Objects.isNull(highScoreTokens) || highScoreTokens.isEmpty()) {
```

Comment:
I think we only need to check the highScoreTokens is null, and set to empty when it's null. We don't need to throw exception when it's empty.

Because we can't exclude the case the model can generate empty sparse vector for input text. Although it's rare, but still possible